### PR TITLE
Determine if one boundary is within other boundary

### PR DESCRIPTION
**Pull request by brggbr**
Issues: #66 and #67 
file location: `SUAS-2023/vision/deskew/within_bounds.py`

`contains_airdrop_boundary(list[list[float]], list[list[float]])`
- Given two lists of points containing x and y coordinates, determines whether any part of either shape created by each list intersects or exists within the other.
- Returns True if any point within the airdrop_boundary is within the image, False otherwise.

`inside_bounds(list[list[float]], list[float])`
- Given one point containing x and y coordinates and a list of points containing x and y coordinates, determines whether the point exists within the boundary created by the list.
- Returns True if the location_point is within the bounds of the shape created by boundary. Returns False otherwise.

### CODE REVIEW

File: vision/deskew/within_bounds.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    ----------
+    point1: [float]
+        The first point on the line segment.
+        An Array that contains the x and y coordinates of the point.
+    point2: [float]
+        A point that may or may not lie on the line segment that is
+        created by point1 and point3.
+        An Array that contains the x and y coordinates of the point.
+    point3: [float]
+        The end point on the line segment.
```
Comment (cjhr95) [inline vision/deskew/within_bounds.py]: The notation here would be:
`point1: list[float]`

File: vision/deskew/within_bounds.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    ----------
+    point1: [float]
+        The first point of a shape or possibly line.
+        An Array that contains the x and y coordinates of the point.
+    point2: [float]
+        The second point of a shape or possibly line.
+        An Array that contains the x and y coordinates of the point.
+    point3: [float]
+        The third point of a shape or possibly line.
+        An Array that contains the x and y coordinates of the point.
```
Comment (cjhr95) [inline vision/deskew/within_bounds.py]: Same as above

File: vision/deskew/within_bounds.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    Parameters
+    ----------
+    line1: [[float]]
+        The first line that may or may not intersect with line2.
+        A 2D array that contains two points and each point is made up of
+        x and y coordinates.
+    line2: [[float]]
+        The second line that may or may not intersect with line1.
+        A 2D array that contains two points and each point is made up of
+        x and y coordinates.
```
Comment (cjhr95) [inline vision/deskew/within_bounds.py]: The nested data structure here would have a similar annotation in the docstring as the function definition:
`line1: [list[list[float]]]`

File: vision/deskew/within_bounds.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    """
+    Returns True if the location_point is within the bounds of the shape
+    created by boundary. Returns False otherwise.
+
+    Parameters
+    ----------
+    boundary: [[float]]
+        Contains a number of points which create a shape.
+        A 2D array which contains a number of points where each point
+        is made up of x and y coordinates.
```
Comment (cjhr95) [inline vision/deskew/within_bounds.py]: Same as above:
`boundary: list[list[float]]`

File: vision/deskew/within_bounds.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    location_point: [float]
```
Comment (cjhr95) [inline vision/deskew/within_bounds.py]: Same as above:
`location_point: list[float]`

File: vision/deskew/within_bounds.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+    Parameters
+    ----------
+    image_corners: [[float]]
+        Contains a number of points which create a shape.
+        A 2D array which contains a number of points where each point
+        is made up of x and y coordinates.
+    airdrop_boundary : [[float]]
+        Contains a number of points which create a shape.
+        A 2D array which contains a number of points where each point
```
Comment (cjhr95) [inline vision/deskew/within_bounds.py]: Same formatting as above

File: vision/deskew/within_bounds.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+    References
+    ----------
+    https://www.geeksforgeeks.org/check-if-two-given-line-segments-intersect/
+    """
+
+    to_return: int = 0  # Defaults to_return as Colinear
+
+    # GeeksforGeeks algorithm
+    val: int = ((point2[1] - point1[1]) * (point3[0] - point2[0])) - (
```
Comment (cjhr95) [inline vision/deskew/within_bounds.py]: The type of var is a float. see the mypy error

**Review by cjhr95 — CHANGES_REQUESTED** [review]
Just some small formatting stuff

File: vision/deskew/within_bounds.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    to_return: bool
+        Returns True if point2 lies on the line segment created by
+        point1 and point2. Returns False otherwise.
+
+    References
+    ----------
+    https://www.geeksforgeeks.org/check-if-two-given-line-segments-intersect/
+    """
+
+    to_return: bool = False
```
Comment (eblake003) [inline vision/deskew/within_bounds.py]: I would rename `to_return` to something like `is_on_line`. Also, you don't need the `if` statement, you can just write `is_on_line: bool = ((point2[0]`...

File: vision/deskew/within_bounds.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+
+def on_line(point1: list[float], point2: list[float], point3: list[float]) -> bool:
+    """
+    Takes three points and returns True if point2 is on the line
+    segment created by point1 and point3. Returns False otherwise.
+
+    Parameters
+    ----------
+    point1: list[float]
```
Comment (eblake003) [inline vision/deskew/within_bounds.py]: Rename parameters to be more clear about their purpose. I would suggest `line_start`, `line_end`, and `test_point`

File: vision/deskew/within_bounds.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    -------
+    to_return: bool
+        Returns True if line1 and line2 intersect and False otherwise.
+
+    References
+    ----------
+    https://www.geeksforgeeks.org/check-if-two-given-line-segments-intersect/
+    """
+
+    to_return: bool = False
```
Comment (eblake003) [inline vision/deskew/within_bounds.py]: Same thing here. Rename this to `intersects` or something similar

File: vision/deskew/within_bounds.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+    # pointPolygonTest returns 1.0 if point is within bounds, and -1.0
+    # if the point is not within bounds
+    if ppt_return > 0:
+        is_in_bounds = True
+
+    return is_in_bounds
+
+
+def contains_airdrop_boundary(
```
Comment (eblake003) [inline vision/deskew/within_bounds.py]: Consider putting the higher level functions at the top (`contains_airdrop_boundary()`, `inside_bounds()`) to make it immediately apparent what the file is for.

File: vision/deskew/within_bounds.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    -------
+    is_in_bounds: bool
+        Returns True if the location_point is within the bounds of the
+        shape created by boundary. Returns False otherwise.
+    """
+
+    is_in_bounds: bool = False
+
+    # Transforms boundary array to numpy array
+    new_np_boundary: NDArray[Shape["*, *"], Float32] = np.array(boundary)
```
Comment (eblake003) [inline vision/deskew/within_bounds.py]: _Technically_ speaking, the `boundary` variable is a list and not an array. If you want slightly better names, you could use `boundary_list` for the list and `boundary_array` for the numpy array.

**Review by eblake003 — CHANGES_REQUESTED** [review]
Looks good! Just a couple small changes just to make things match my own personal preference

File: vision/deskew/within_bounds.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+    boundary_array = boundary_array.reshape((-1, 1, 2))
+
+    # Tests if point is within boundary
+    ppt_return: float = cv2.pointPolygonTest(
+        boundary_array, (location_point[0], location_point[1]), False
+    )
+
+    # pointPolygonTest returns 1.0 if point is within bounds, and -1.0
+    # if the point is not within bounds
+    if ppt_return > 0:
```
Comment (eblake003) [inline vision/deskew/within_bounds.py]: `is_in_bounds` can just be set equal for `ppt_return > 0`

File: vision/deskew/within_bounds.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+    -------
+    intersects: bool
+        Returns True if line1 and line2 intersect and False otherwise.
+
+    References
+    ----------
+    https://www.geeksforgeeks.org/check-if-two-given-line-segments-intersect/
+    """
+
+    intersects: bool = False
```
Comment (eblake003) [inline vision/deskew/within_bounds.py]: Move this down to right before the if statements

**Review by eblake003 — CHANGES_REQUESTED** [review]
Couple small changes

**Review by eblake003 — APPROVED** [review]
Looks good!

**Review by cjhr95 — APPROVED** [review]
Nice work!

**Pull request merged**